### ISSUE TITLE
Make SolutionArray picklable

### ISSUE THREAD

**Issue by ischoegl** [Cantera/enhancements#162]
**Abstract**

After refactoring of `SolutionArray` in Cantera/cantera#1426 (and Cantera/cantera#1458), there is still the option of making the object picklable in Python (similar to Cantera/cantera#692). An implementation is deferred, as it is unclear how common use cases would be. 

The implementation itself would be straight-forward: modify C++ `SolutionArray::save/restore` to allow for YAML string output/input, and implement Python `SolutionArray.__getstate__/__getstate__`.

**Comment by rwest** [comment]
In general having things be  pickleable is nice, especially for some parallelization and workflow tools that rely on it

**Comment by ischoegl** [comment]
> In general having things be pickleable is nice, especially for some parallelization and workflow tools that rely on it

Thanks for the comment. I generally agree, and have implemented Cantera/cantera#692 for that reason. Pickled `SolutionArray` objects may be a good way to pass back results from `multiprocessing` pools. I don’t have an urgent need for this at the moment myself, but if anyone else wants to take this on I’d be more than happy to assist with a PR.

**Comment by mruijzendaal** [comment]
Seconding the request for a picklable `SolutionArray`. I have no experience with the C++ interfacing, so I'm afraid I can't contribute to a PR.

I hacked this workaround in `cantera/composite.py`, replacing the placeholder of `__reduce__`. Perhaps this will be of use to other seeking to use pickle with `SolutionArray`:

```python diff
class SolutionArray(SolutionArrayBase):
    ...
    def __reduce__(self):
        import os
        from pathlib import Path

        # Save to a temporary YAML file
        tf = Path("cantera_solarr_save_tempfile.yaml")
        if (tf.exists()):
            os.unlink(tf)
        self.save(tf.name, name="solarr", overwrite=True)

        # Read the YAML file into a string and delete the file
        with open(tf, 'r') as f:
            yaml_data = f.read()
        os.unlink(tf.name)

        state = {
            'yaml_data': yaml_data,
            'phase': self._phase, # Solution object, which is already picklable
        }
        return (self.__class__._from_pickle, (state,))

    @classmethod
    def _from_pickle(cls, state):
        from pathlib import Path
        import os
        # Recreate phase
        phase = state.get('phase')
        
        # Create empty SolutionArray
        arr = cls(phase)

        # Write the YAML string to a file
        tf = Path("cantera_solarr_load_tempfile.yaml")
        if (tf.exists()):
            os.unlink(tf)
        with tf.open('w') as f:
            f.write(state['yaml_data'])
            f.flush()
        # Restore the SolutionArray from the yaml
        arr.restore(tf.name, 'solarr')

        # Clean up the temporary file
        os.unlink(tf.name)

        return arr
```

This works, but is kind of dirty. The approach mentioned above (getting the YAML string from C++ directly) would be preferable.

**Comment by ischoegl** [comment]
Thanks, @mruijzendaal ... this is indeed a workable implementation. Outputting/inputting strings into C++ directly may be more elegant, but your approach to using temporary files is a great proof-of-concept. The resulting pickled object already covers the most straightforward implementation.

**Comment by mruijzendaal** [comment]
If you prefer, I could open a pull request implementing this change. If not, we can leave this for any others needing this feature.

**Comment by ischoegl** [comment]
> If you prefer, I could open a pull request implementing this change. If not, we can leave this for any others needing this feature.

While I cannot speak for all maintainers, I’d be :+1: with this becoming a PR. Beyond, I wonder whether this may be workable as a file holding YAML: https://docs.python.org/3/library/tempfile.html#module-tempfile

**Comment by mruijzendaal** [comment]
> While I cannot speak for all maintainers, I’d be 👍 with this becoming a PR. Beyond, I wonder whether this may be workable as a file holding YAML: https://docs.python.org/3/library/tempfile.html#module-tempfile

Good pointer. `tempfile.TemporaryFile()` doesn't work since it outputs a file object, while `SolutionArray.save()` appears to need a file path string. I worked out a solution with `tempfile.TemporaryDirectory()` that works and cleans up after itself. As a bonus, I also implemented `__copy__` which was also marked with `NotImplementedError` and can use a similar syntax.

```python
class SolutionArray(SolutionArrayBase):
    ...
    def _to_picklable(self):
        import os
        from pathlib import Path
        import tempfile

        # Save to a temporary YAML file
        with tempfile.TemporaryDirectory() as tmpdir:
            tf = Path(tmpdir, "cantera_solarr_save_tempfile.yaml")

            # Remove the file if it already exists
            if (tf.exists()):
                os.unlink(tf)
            self.save(tf, name="solarr", overwrite=True)

            # Read the YAML file into a string and delete the file
            with open(tf, 'r') as f:
                yaml_data = f.read()

        return {
            'yaml_data': yaml_data,
            'phase': self._phase, # Solution object, which is already picklable
        }

    @classmethod
    def _from_pickle(cls, state):
        from pathlib import Path
        import os
        import tempfile
        # Recreate phase
        phase = state.get('phase')
        
        # Create empty SolutionArray
        arr = cls(phase)

        # Write the YAML string to a temporary file
        with tempfile.TemporaryDirectory() as tmpdir:
            tf = Path(tmpdir, "cantera_solarr_load_tempfile.yaml")

            # Remove the file if it already exists
            if (tf.exists()):
                os.unlink(tf)
            with tf.open('w') as f:
                f.write(state['yaml_data'])
                f.flush()
                
            # Restore the SolutionArray from the yaml
            arr.restore(tf, 'solarr')

        return arr
    
    def __reduce__(self):
        return (self.__class__._from_pickle, (self._to_picklable(),))

    def __copy__(self):
        return self.__class__._from_pickle(self._to_picklable())
```

I'll open up a pull request to see what other maintainers think about this.

**Comment by ischoegl** [comment]
@mruijzendaal … great! Have you looked into `tempfile.NamedTemporaryFile`? (Haven’t used that myself, but it may work.) Regardless, I am looking forward to a PR. If possible, please include a test - there are some [rudimentary instructions](https://cantera.org/dev/develop/writing-tests.html) but we can always provide additional pointers as necessary.

**Comment by mruijzendaal** [comment]
See [cantera/pull/1964](https://github.com/Cantera/cantera/pull/1964). Thanks for the cooperation so far!

**Comment by mruijzendaal** [comment]
> [@mruijzendaal](https://github.com/mruijzendaal) … great! Have you looked into `tempfile.NamedTemporaryFile`? (Haven’t used that myself, but it may work.) Regardless, I am looking forward to a PR. If possible, please include a test - there are some [rudimentary instructions](https://cantera.org/dev/develop/writing-tests.html) but we can always provide additional pointers as necessary.

Yes, unfortunately `tempfile.NamedTemporaryFile` had the same result for me. 

I included a basic test. Not crashing seems like the primary test purpose, since `Solution` and writing YAMLs are tested elsewhere.

**Comment by ischoegl** [comment]
Closed via Cantera/cantera#1964
